src/main/java/me/jellysquid/mods/lithium/mixin/entity/item_self_insert/HopperBlockEntityMixin.java

```diff
@@ -0,0 +1,72 @@
+package me.jellysquid.mods.lithium.mixin.entity.item_self_insert;
+
+import me.jellysquid.mods.lithium.common.blockentities.HopperAccess;
+import net.minecraft.block.HopperBlock;
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.block.entity.BlockEntityType;
+import net.minecraft.block.entity.Hopper;
+import net.minecraft.block.entity.HopperBlockEntity;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.function.Supplier;
+
+@Mixin(HopperBlockEntity.class)
+public abstract class HopperBlockEntityMixin extends BlockEntity implements HopperAccess {
+    // this represents the number of ticks until the hopper is ready to accept another item
+    private int realCooldown;
+
+    @Shadow
+    protected abstract boolean isFull();
+
+    @Shadow
+    private int transferCooldown;
+
+    public HopperBlockEntityMixin(BlockEntityType<?> type) {
+        super(type);
+    }
+
+    @Inject(method = "extract(Lnet/minecraft/block/entity/Hopper;)Z", cancellable = true, at = @At(value = "INVOKE", target = "Lnet/minecraft/block/entity/HopperBlockEntity;getInputItemEntities(Lnet/minecraft/block/entity/Hopper;)Ljava/util/List;"))
+    private static void extract(Hopper hopper, CallbackInfoReturnable<Boolean> cir) {
+        if (hopper instanceof HopperBlockEntity && !((HopperBlockEntityMixin) hopper).full()) {
```

Comment:
This cast shall fail as the `HopperBlockEntity` won't be an instance of `HopperBlockEntityMixin` at runtime. Consider exposing the method to `HopperAccess` instead

Reply:
Oops